### PR TITLE
Implement reattach process using stored dataset

### LINKED ISSUE #867: Implement reattach process using stored dataset
The active dataset and pending dataset could be stored in non-volatile memory with #866, then we need to implement reattach process after reset, and it will help to pass certification test 9.2.8, 9.2.15, 9.2.16.

### PR DESCRIPTION
**Pull request by chshu**
Resolves #867.

This PR implements the reattach process using stored dataset:
1. Reattach process will only be enabled with command "thread start";
2. It will try to attach using stored active dataset firstly, if there is no active dataset, stop reattach process;
3. If it failed to attach using active dataset, it will apply pending dataset and try to attach again, if there is no pending dataset, stop reattach process;
4. If it failed to attach using pending dataset, it will revert back to using the active dataset, and enable local Pending dataset and start its delay timer.

It helps to pass certification test: ED_9_2_8 and Router_9_2_8:
[Router_9_2_8.zip](https://github.com/openthread/openthread/files/555820/Router_9_2_8.zip)
[ED_9_2_8.zip](https://github.com/openthread/openthread/files/555821/ED_9_2_8.zip)

While other three related tests still failed due to some other issues, I will go on to analyze and fix them:
[SED_9_2_8.zip](https://github.com/openthread/openthread/files/555972/SED_9_2_8.zip)
[Router_9_2_15.zip](https://github.com/openthread/openthread/files/555856/Router_9_2_15.zip)
[Router_9_2_16.zip](https://github.com/openthread/openthread/files/555857/Router_9_2_16.zip)

### CODE REVIEW

File: src/core/thread/mle.hpp [hunk 1]
```diff
@@ -1160,6 +1162,19 @@ class Mle
...
+    /**
+     * States when reattaching network using stored dataset
+     *
+     */
+    enum ReattachState
+    {
+        kReattachStart      = 1,   ///< Start reattach process
+        kReattachActive     = 2,   ///< Reattach using stored Active Dataset
+        kReattachPending    = 3,   ///< Reattach using stored Pending Dataset
+        kReattachStop       = 4,   ///< Reattach process finished
```
Comment (jwhui) [inline src/core/thread/mle.hpp]: nit: prefer assigning default state `kReattachStop` to `0`.

File: src/core/thread/mle.cpp [hunk 2]
```diff
@@ -246,6 +252,7 @@ ThreadError Mle::Stop(void)
         mNetif.RemoveUnicastAddress(mLeaderAloc);
     }
 
+    mNetif.GetPendingDataset().StopDelayTimer();
```
Comment (jwhui) [inline src/core/thread/mle.cpp]: Where do we restart the Delay Timer?  If we detach and we have a Pending Dataset, we need to make sure that we apply the Pending Dataset after the Delay Timer expires.

File: src/core/thread/mle.cpp [hunk 3]
```diff
@@ -1157,37 +1176,61 @@ void Mle::HandleParentRequestTimer(void)
...
-                    BecomeDetached();
+                    mReattachState = kReattachStop;
                 }
-                else if (mMleRouter.BecomeLeader() != kThreadError_None)
+            }
+            else if (mReattachState == kReattachPending)
+            {
+                mReattachState = kReattachStop;
+                mNetif.GetActiveDataset().Restore();
+                mNetif.GetPendingDataset().Set(mNetif.GetPendingDataset().GetLocal());
```
Comment (jwhui) [inline src/core/thread/mle.cpp]: What is the purpose of setting the Pending Dataset here?  To restart the Delay Timer?

File: src/core/thread/mle.hpp (hunk 1 above)
Comment (chshu) [inline src/core/thread/mle.hpp]: Updated, thanks.

File: src/core/thread/mle.cpp (hunk 2 above)
Comment (jwhui) [inline src/core/thread/mle.cpp]: Do we have all cases covered?  What about the case where we successfully attach but detach before properly receiving the Active/Pending Datasets from the Parent.  Will we restart the Delay Timer in this case?

Comment (chshu) [inline src/core/thread/mle.cpp]: Yes, this case is a problem, we need to restart the Delay Timer if it attach successfully without refreshing network Pending dataset.

I have added the logic to check Delay Timer state once it attach/reattach/re-sync to the network (call `IsDelayTimerRunning()` in `setStateChild()` and `setStateRouter()`), and restart the Delay Timer if it is necessary. Please have a review again, thanks.

Comment (jwhui) [inline src/core/thread/mle.cpp]: In general, when a device is attached to a network, it should always use the Delay Timer that was received from its Parent (i.e. the one in `mNetwork`).  If the device detaches from the network, then it should use the one it has stored locally (i.e. the one in `mLocal`).  So rather than starting the Delay Timer in `SetStateChild()` and `SetStateRouter()`, I think it's better if we just start the timer when we are detached.  Does that make sense?

Comment (chshu) [inline src/core/thread/mle.cpp]: Here the Delay Timer is stopped in `Mle::Stop()` which force the device to be detached, I think it's not a normal detach, we can't control the time period from next 'thread start' and how many network changes may happen during the period, so I think it is not necessary to start mLocal's Delay Timer here, just try to store current datasets and wait for next `thread start`;

And for Certification test, I don't want the Active Dataset be overwritten by Pending Dataset after 'thread stop', so we could have both Active Dataset and Pending Dataset stored before next 'thread start' and start the reattach process with them.

At the same time, although the Delay Timer is stopped here, we still can make sure the Pending Dataset will take effect in all cases after "thread start":
- If device re-sync to network with `Child Synchronization`/`Link Synchronization` process successfully, Delay Timer will be restarted in `SetStateChild()`/`SetStateRouter()`;
- If device fails to re-sync, it will start reattach process with Active Dataset firstly, if successfully attach, Delay Timer will be restarted in `SetStateChild()`/`SetStateRouter()`;
- If device fails to attach using Active Dataset, it will try to attach using Pending Dataset, so Pending Dataset takes effect here;
- If device fails to attach using Pending Dataset, it will revert back to using Active Dataset and start Delay Timer for Pending Dataset, Pending Dataset will take effect finally.

Please correct me if I missed something, thanks a lot for your time and patient.

Comment (jwhui) [inline src/core/thread/mle.cpp]: The scenario I'm trying to address is when a device is attached to the network, it should be in sync with the network.  In other words, if the device has a Pending Dataset that is not synchronized with the network, it should try to remain attached to that network so that it can provide the Pending Dataset to the Leader.  Starting the Delay Timer in `SetStateChild()` and `SetStateRouter()` can cause the device to apply its Pending Dataset before it's had a chance to synchronize it with the rest of the network.

Comment (jwhui) [inline src/core/thread/mle.cpp]: Maybe I'm missing something, but `thread stop` will cause the device to detach from the network and can then manage its Delay Timer based on its local pending dataset.  My previous point is that we should restart the delay timer when:

1. the device has attached to a network and received a Pending Dataset from its Parent.
2. the device has detached from the network on its own (i.e. due to loss of connectivity).

I understand that we currently need to stop the Delay Timer as part of `Mle::Stop()` and that is fine for now.  `Mle::Stop()` is only called as part of `thread stop`, so it's fine to stop the Delay Timer in there for now.  But we should also consider restarting the Delay Timer if the device has detached from the network on its own (i.e. due to losing connectivity with its Parent).  Remember that there are two different reasons for detaching from the network: (i) the device detaches due to calling `thread stop` and (ii) the device detaches due to loss of connectivity.  Right now, you are attempting to handle (i).  I think we can also handle (ii) properly.

Comment (chshu) [inline src/core/thread/mle.cpp]: Thanks a lot for your clarification.
For (i), I have updated the PR to stop the Delay Timer only if `Mle::Stop()` is triggered by `thread stop`;
For (ii), If the device has detached from the network on its own, we need to stop the current mNetwork's Delay Timer, and restart a new Timer for mLocal. I want to use another separate PR to address it, since it seems a separate feature enhancement, Do you agree?

Comment (jwhui) [inline src/core/thread/mle.cpp]: I think it's fine for (ii) to be in a separate PR.  Should we also remove starting the Delay Timer in `SetStateChild()` and `SetStateRouter()`?  The timer should be started after receiving the Pending Dataset from its Parent.

Comment (chshu) [inline src/core/thread/mle.cpp]: The starting Delay Timer action in `SetStateChild()` and `SetStateRouter()` is to address the following case: the device detached due to `thread stop` command, while the network datasets are still exist, and then after calling `thread start`, the device will re-sync/reattach to network without refreshing Pending Dataset since it already have the same network datasets, this way the timer is not started, so I check the Delay Timer state here and restart it if it is not running.
I have also consider to clear network datasets (`mNetwork`) in `Mle::Stop()`, to ensure the device could refresh its network datasets during reattach process, so then we don't need to check and restart Delay Timer in `SetStateChild()` and `SetStateRouter()`, do you think this way is better?

Comment (jwhui) [inline src/core/thread/mle.cpp]: Yes, I think that's better.  I think this will bring the code closer to when we perform a full reboot of the node.

Comment (chshu) [inline src/core/thread/mle.cpp]: Yes, exactly. I have updated the PR by clearing the network datasets in `Mle::Stop()`(only for `thread start`) instead of stopping Delay Timer. Thanks.
